### PR TITLE
Capture character distribution within strings in SSTs

### PR DESCRIPTION
**Pull request by wemrysi**
Extends the structural treatment of strings in SSTs to widen strings shorter than `stringMaxLength` to an array of their characters allowing the preservation of positional character distributions. This enables SSTs to capture much of the structure that may be embedded in a string (i.e. SSNs, phone numbers, UUIDs, temporal, etc).

#qz-3719 Done.

### CODE REVIEW

**Comment by alissapajer** [comment]
Cool!

File: web/src/test/scala/quasar/api/services/analyze/SchemaServiceSpec.scala [hunk 1]
```diff
@@ -128,17 +135,33 @@ final class SchemaServiceSpec extends quasar.Qspec with FileSystemFixture with H
...
+      }
 
-        service(InMemState.empty)(Request(uri = ruri))
-          .flatMap(_.as[ApiError])
-          .map(_ must beApiErrorWithMessage(
-            BadRequest withReason "Invalid query parameter."))
-          .unsafePerformSync
+    val nonPositiveInvalid =
+      List("unionMaxSize") map { param =>
+        s"non-positive $param given" >> prop { (file: AFile, i: Int) =>
```
[the remark this answers is not in the document]
Comment (wemrysi) [reply web/src/test/scala/quasar/api/services/analyze/SchemaServiceSpec.scala]: It doesn't make sense for a type union to have zero members, the smallest it could become would be `Top`.

[the remark this answers is not in the document]
Comment (alissapajer) [reply web/src/test/scala/quasar/api/services/analyze/SchemaServiceSpec.scala]: Ah, it's type union. Got it. :+1:

File: frontend/src/main/scala/quasar/sst/strings.scala [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      }
+
+    stringTagged(strStat, C.embed(envT(strStat, TypeST(TypeF.arr(charArr)))))
+  }
+
+  /** Widens a string into an array of its characters.
+    *
+    * FIXME: Overly specific, define in terms of [Co]Recursive.
+    */
+  def widen[J: Order, A: ConvertableTo: Field: Order](count: A, s: String)(
```
[the remark this answers is not in the document]
Comment (wemrysi) [reply frontend/src/main/scala/quasar/sst/strings.scala]: The count in `widen` is the number of observations of the input string, this is required as we're producing an `SST` from the function.

The `String` argument provided to `compress` _is_ the string to compress, which is achieved by accumulating stats for the characters within the string.

I'll add a couple of specific tests for these functions.

File: frontend/src/test/scala/quasar/sst/CompressionSpec.scala [hunk 3]
```diff
@@ -262,26 +274,32 @@ final class CompressionSpec extends quasar.Qspec
...
-      val arr  = envT(ts, TypeST(TypeF.arr[J, S](char.right))).embed
+      val str = SST.fromEJson(Real(1), J.str(s))
+      val arr = strings.compress[S, J, Real](str.copoint, s).embed
 
-      val req = str.transCata[S](compression.limitStrings(plen))
-      val rlt = str.transCata[S](compression.limitStrings(lt))
+      val req = attemptCompression(str, compression.limitStrings(plen))
+      val rlt = attemptCompression(str, compression.limitStrings(lt))
 
       (req must_= str) and (rlt must_= arr)
```
[the remark this answers is not in the document]
Comment (wemrysi) [reply frontend/src/test/scala/quasar/sst/CompressionSpec.scala]: It's testing boundary conditions, `L301` shows that a string having length equal to `maxLen` is _not_ compressed and `L302` shows that a string having a length one greater than `maxLen` _is_ compressed.

File: README.md [hunk 4]
```diff
@@ -739,17 +740,33 @@ a schema document might look like
     "of" : {
       "city" : {
         "measure" : {
+          "kind" : "collection",
           "count" : 1000.0,
-          "min" : "ABBEVILLE",
-          "max" : "YOUNGSVILLE",
```
Comment (alissapajer) [inline README.md]: We need to coordinate with the frontend on this schema change (not specifically this line but the expected schema as a whole).

@beckyconning is the product currently using schema in any ways that this schema change would break?

**Comment by beckyconning** [comment]
We aren't currently using SSTs in the frontend so :+1:
